src/main/java/com/puppycrawl/tools/checkstyle/xpath/AbstractNode.java

```diff
@@ -49,6 +49,9 @@
     /** The children. */
     private final List<AbstractNode> children = new ArrayList<>();
 
+    /** The {@code TreeInfo} object. */
+    private TreeInfo treeInfo;
```

Comment:
`TreeInfo` could be `final` if constructor was made to accept it as a parameter, then all sub-classes would just have to give it through `super` call.

Reply:
Done